### ISSUE TITLE
multiple jaxb-api jars can cause NullPointerException in ContextFinder

### ISSUE THREAD

**Issue by Tomas-Kraus** [eclipse-ee4j/jaxb-ri#863]
Unrelated to this bug, I have a problem in my app with multiple copies of the jaxb-api in the JVM at once. This leads to a class cast exception in ContextFinder.find(). More about that below.

The immediate problem is that this leads to a very confusing NullPointerException in the error reporting code here:

private static JAXBException handleClassCastException(Class originalType, Class targetType)

{ final URL targetTypeURL = which(targetType); return new JAXBException(Messages.format(Messages.ILLEGAL_CAST, // we don't care where the impl class is, we want to know where JAXBContext lives in the impl // class' ClassLoader originalType.getClassLoader().getResource("javax/xml/bind/JAXBContext.class"), targetTypeURL)); }

The problem is that originalType is "com.sun.xml.internal.bind.v2.runtime.JAXBContextImpl" (from rt.jar in the JDK). getClassLoader() on that type returns null (indicating the bootloader)

Now the background: the JRE has v2.0 of javax.xml.bind, and my OSGi environment has v2.2 of jaxb-api via ServiceMix. My code is invoking the v2.2 version via JAXBContext.newInstance(), but that code finds the context factory from the JRE, so it instantiates com.sun.xml.internal.bind.v2.runtime.JAXBContextImpl which implements the JAXBContext from the JRE instead of the JAXBContext from ServiceMix, thus the class cast exception.

Of course, if I solve my multiple-jar problem, then the null pointer will go away. But that NPE made it REALLY hard to figure out what was actually happening. Hence this bug report.
#### Environment
OSGi 4.2, Karaf 2.2.2, Talend Service Factory 2.4.2.0, Win7 x64, JDK 1.6.0_26
#### Affected Versions
[2.2]

**Comment by perlun** [comment]
I can confirm that this is still a bug with `jakarta.xml.bind-api-3.0.0.jar`. The problem seems to be that `getClassClassLoader(originalType)` is blindly presumed to have returned a non-`null` reference => causing the NPE.

![image](https://user-images.githubusercontent.com/630613/153152215-53191f15-2042-40e8-91cc-60c524ae167c.png)

Digging a bit deeper, this seems to be caused by the `c.getClassLoader()` method returning `null`:

![image](https://user-images.githubusercontent.com/630613/153153006-d1a59f3a-30e2-40aa-9d36-57d32226ff37.png)

I hope this helps, in case someone decided to fix the error handling to fail more gracefully. :+1:

**Comment by lukasj** [comment]
`...internal...JAXBContextImpl` is not an instance of `jakarta.xml.bind.JAXBContext`, the issue in your case is elsewhere

**Comment by perlun** [comment]
> `...internal...JAXBContextImpl` is not an instance of `jakarta.xml.bind.JAXBContext`, the issue in your case is elsewhere

Absolutely, I'm not in any way claiming that this should "work". My point is though that the `handleClassCastException` should ideally handle these cases more gracefully. In other words, the following method:

```java
    private static JAXBException handleClassCastException(Class originalType, Class targetType) {
        final URL targetTypeURL = which(targetType);

        return new JAXBException(Messages.format(Messages.ILLEGAL_CAST,
                // we don't care where the impl class is, we want to know where JAXBContext lives in the impl
                // class' ClassLoader
                getClassClassLoader(originalType).getResource("jakarta/xml/bind/JAXBContext.class"),
                targetTypeURL));
    }
```

...should perhaps be modified to something more like this. Or do you see any problems with such an approach?

```java
    private static JAXBException handleClassCastException(Class originalType, Class targetType) {
        final URL targetTypeURL = which(targetType);

        ClassLoader originalTypeClassLoader = getClassClassLoader(originalType);

        if (originalTypeClassLoader != null) {
            return new JAXBException(Messages.format(Messages.ILLEGAL_CAST,
                    // we don't care where the impl class is, we want to know where JAXBContext lives in the impl
                    // class' ClassLoader
                    originalTypeClassLoader.getResource("jakarta/xml/bind/JAXBContext.class"),
                    targetTypeURL));
        }
        else {
            // TODO: create and return another JAXBException in this case, failing gracefully.
        }
```
